node/core/approval-voting/src/lib.rs

```diff
@@ -539,6 +534,117 @@ struct ApprovalStatus {
 	block_tick: Tick,
 }
 
+#[derive(Copy, Clone)]
+enum ApprovalOutcome {
+	Approved,
+	Failed,
+	TimedOut,
+}
+
+struct ApprovalState {
+	validator_index: ValidatorIndex,
+	candidate_hash: CandidateHash,
+	candidate_index: CandidateIndex,
+	approval_outcome: ApprovalOutcome,
+}
+
+impl ApprovalState {
+	fn approved(
+		validator_index: ValidatorIndex,
+		candidate_hash: CandidateHash,
+		candidate_index: CandidateIndex,
+	) -> Self {
+		Self {
+			validator_index,
+			candidate_hash,
+			candidate_index,
+			approval_outcome: ApprovalOutcome::Approved,
+		}
+	}
+	fn failed(
+		validator_index: ValidatorIndex,
+		candidate_hash: CandidateHash,
+		candidate_index: CandidateIndex,
+	) -> Self {
+		Self {
+			validator_index,
+			candidate_hash,
+			candidate_index,
+			approval_outcome: ApprovalOutcome::Failed,
+		}
+	}
+}
+
+struct CurrentlyCheckingSet {
+	candidate_hash_map: HashMap<CandidateHash, Vec<Hash>>,
+	currently_checking: FuturesUnordered<BoxFuture<'static, ApprovalState>>,
+}
+
+impl Default for CurrentlyCheckingSet {
+	fn default() -> Self {
+		Self {
+			candidate_hash_map: HashMap::new(),
+			currently_checking: FuturesUnordered::new(),
+		}
+	}
+}
+
+impl CurrentlyCheckingSet {
+	// The return value of this function will indicate whether we have seen this
+	// candidate_hash for this relay_block before. If we have never seen this candidate_hash
+	// xor never seen this relay_block (for this candidate_hash), the function will return false
```

Comment:
Comment here is outdated; should instead indicate that the `launch-work` future is only used if the candidate is not already undergoing validation